### PR TITLE
 ARM64 Pagerando optimizer

### PR DESCRIPTION
**Pull request by yln**
And some cleanup changes to the ARM32 optimizer

### CODE REVIEW

File: lib/CodeGen/PagerandoBinning.cpp [hunk 1]
```diff
@@ -8,14 +8,16 @@
...
+// (and pagerando wrappers) are not assigned to a bin.
 // Function sizes are estimated by adding up the size of all instructions
 // inside the corresponding MachineFunction. The default bin size is 4KB.
 // The current bin allocation strategy is a greedy algorithm that, for every
 // function, picks the bin with the smallest remaining free space that still
 // accommodates the function. If such a bin does not exist, a new one is
 // created. Functions that are larger than the default bin size are assigned to
 // a new bin which forces the expansion of said bin.
+// Note: Because this pass estimates function sizes it should run as late as
+// possible.
```
Comment (rinon) [inline lib/CodeGen/PagerandoBinning.cpp:20]: We should note that it does require machine SSA so currently needs to be scheduled pre-register allocation. Register allocation breaks easy use-def tracking.

**Review by rinon — COMMENTED** [review]
LGTM

File: lib/CodeGen/PagerandoBinning.cpp (hunk 1 above)
Comment (yln) [reply lib/CodeGen/PagerandoBinning.cpp:20]: You mean the optimizer right? This is binning.
I think for the optimizer it's explicitly stated in the code. Maybe even enforced?
```cpp
MachineFunctionProperties getRequiredProperties() const override {
    return MachineFunctionProperties().set(
        MachineFunctionProperties::Property::TracksLiveness);
  }
```

Comment (rinon) [reply lib/CodeGen/PagerandoBinning.cpp:20]: Still probably worth a comment, to bound "as late as possible."

Comment (yln) [reply lib/CodeGen/PagerandoBinning.cpp:20]: Added a comment in my new branch/next PR.

**Pull request merged**